docs/contributing/plugins.md

```diff
@@ -0,0 +1,63 @@
+│   ├── <other source files or folder>
+├── <other files or folder>
+```
+
+Trestle uses naming convention to discover the top-level module of the plugin projects. It expects the top-level module to be named `trestle_{plugin_name}`. In the above example, the top-level module is named as `trestle_fedramp` so that it can be autmatically discovered by trestle. All the python source files should be created inside this module (folder).
+
+The top-evel module should contain a `commands` directory where all the plugin command files should be stored. Each command should have its own python file. In the above exaample, `validate.py` file conatins one command for this plugin. Other python files or folders should be created in the top-level module folder, outside the `commands` folder. This helps in keeping the commands separate and in their discovery by trestle.
+
+## Command Creation
+
+The plugin command should be created as shown in the below code snippet.
+
+```
+from trestle.core.commands.command_docs import CommandBase
+from trestle.core.commands.command_docs import CommandPlusDocs
+
+class ValidateCmd(CommandBase):
+    """Validate contents of an OSCAL model based on FedRAMP specifications."""
+
+    name = 'fedramp-validate'
+
+    def _init_arguments(self) -> None:
+        logger.debug('Init arguments')
+        self.add_argument('-f', '--file', help='OSCAL file to validate.', type=str, required=True)
+
+        self.add_argument(
+            '-o', '--output-dir', help='Output directory for validation results.', type=str, required=True
+        )
+
+    def _run(self, args: argparse.Namespace) -> int:
+        model_file = pathlib.Path(args.file).resolve()
+
+        output_dir = pathlib.Path(args.output_dir).resolve()
+        ...
+        ...
+        return 0 if valid else 1
+
+```
+
+There should be a command class for example, `ValidateCmd` which should either extend from `CommandBase` or `CommandPlusDocs`. Trestle uses `ilcli` package to create commands. `CommandBase` extends from `ilcli.Command` that initializes the command including help messages and input parameters. `CommandPlusDocs` in turn extends from `CommandBase`. The difference between `CommandBase` and `CommandPLusDocs` is that `CommandBase` does not require command line parameter `trestle-root` to be set or the current directory to be a valid trestle root, whereas `CommandPlusDocs` requires a valid `trestle-root` and checks for it. Hence, depending upon the requirement of the plugin command it can extend from either of these classes.
+
+Input arguments to the command should be specified in `_init_arguments` method as shown above. The acutal code of the command is contained in`_run` method. This method is called by ilcli when the command is excuted on the commandline. The command arguments can be accessed from the `args` input parameter as shown above. The command should return `0` in case of successful execution, or any number greater than 0 in case of failure.
```

Comment:
Return codes should be as `trestle/core/commands/common/return_codes.py`

Reply:
done